kuma/users/tests/test_stripe_subscription.py

```diff
@@ -0,0 +1,42 @@
+from datetime import datetime
+
+import pytest
+
+from django.conf import settings
+from django.test import Client
+from django.urls import reverse
+
+from kuma.core.utils import safer_pyquery as pq
+from kuma.users.models import User
+
+
+@pytest.fixture
+def payments_user(db, django_user_model):
+    return User.objects.create(
+        username='payments_user',
+        email='payer@example.com',
+        date_joined=datetime(2019, 1, 17, 15, 42))
+
+
+def test_create_stripe_subscription(payments_user):
+    if not settings.STRIPE_PLAN_ID:
```

Comment:
well that test only really works with a valid `STRIPE_PLAN_ID` set and I don't wanna put stripe keys into CVS. This basically makes sure that tests still succeeds for PRs where there are none set. I did put stripe keys into travis though, so those tests should run on master.

Reply:
No, this is the wrong approach. Tests should never depend on anything real. At least extremely rarely. Also, you should be able to run the tests without an internet connection every time. 

What you should do is to put the `STRIPE_PLAN_ID` into `settings/testing.py` which is kinda like a fixture. Any persons content of `.env` should *not* cause test execution to be different. 

What I prefer generally, is to put something like...
```python
# in settings/testing.py
SOME_API_KEY_THING = ''
```
Then, I know that 100% of every single test on every computer will always get the same value when the tests start. Same on yours and mine and Travis and Jenkins. 
Now, if you have a test that wants to test that this new API key thing works, you enable explicitly within the test. E.g.:

```python
def test_something(client, settings):
    settings.API_KEY_THING = 'fakefakefake'
    r = client.get(do_something())
```
or, if it's a class-based Django TestCase you have to use a context manager:
```python
    def test_something(self):
        with self.settings(API_KEY_THING='fakefakefake'):
            r = self.client.get(do_something())
```

Reply:
Check out https://github.com/mdn/kuma/pull/6311/files which does this. 
So the Google Analytics event tracking sending is set up to only run if `settings.GOOGLE_ANALYTICS_ACCOUNT` is truthy. [Which it isn't](https://github.com/mdn/kuma/blob/daa553a41d06b18b418ca70ffcf4808b57b3e835/kuma/settings/testing.py#L80).

So, when I need to kick in [I explicitly set it to something](https://github.com/mdn/kuma/blob/daa553a41d06b18b418ca70ffcf4808b57b3e835/kuma/core/tests/test_ga_tracking.py#L8). [And here for class-based tests](https://github.com/mdn/kuma/blob/daa553a41d06b18b418ca70ffcf4808b57b3e835/kuma/users/tests/test_views.py#L1262)).

Also, because I don't trust the vastness of the Kuma test cases, I also took an extra precaution and set `GOOGLE_ANALYTICS_TRACKING_URL` in `settings/testing.py` to something that is unrealistic. If tests were to accidentally fail the mocking and prepping and stuff, it will crash because the domain I put it won't be able to work. (It's not perfect because there's some error swallowing going on here too).

Reply:
Thanks for giving me your perspective on this! Unfortunately I'm not sure if there's a way to meaningful test this PR without the majority of this being a stripe mock. Do you think there's still value in writing tests for this then?

Reply:
Pardon my ignorance but, given the current state of this branch, if I set a `STRIPE_PLAN_ID` in *my* `.env` file, and run the tests it will go out on the internet and talk to Stripe's API server for realz? And if I was in a tunnel or airplane, the test would fail? Or if the network is unreliable at that particular time, it'd hang there and eventually fail even though what I was working on was a typo in the CSS :)

Then, yes I think you should definitely mock out `stripe`. 

There might be room in the `headless` tests, whereever they are, to fully end-to-end test the full HTTP integration with Stripe given a known debug `STRIPE_PLAN_ID`.

Reply:
Yes, that's correct, currently that test would require an active internet connection. I could put a check in there that makes sure the test only fails (and otherwise warns) if one is online. But if we already have that headless directory I'd gladly move it there. I'd like that over mocking the stripe endpoints.
Though, as much as I believe in reproducibility, I don't think we should have a known `STRIPE_PLAN_ID` committed to our repo (as that would also require exposing the sensitive secret key). Instead I'd add documentation to the test case, explaining the preconditions.

Reply:
To be honest, I don't yet fully understand the context of our `headless` tests. You might want to ask @escattone first. 
Either way, these "Django unit tests" should be void of real actual values and should never need an internet connection.